### ISSUE TITLE
Fuzzy queries cause excessive network requests.

### ISSUE THREAD

**Issue by ehuss** [rust-lang/cargo#11934]
When cargo can't find a crate name in a registry, it will also query every permutation of `-` and `_` replacements (up to 1024 times) to see if there is a different match. When this code was written, it was assumed it would just be a simple query to git, but now that the sparse index does a network round trip for each one, it probably isn't a good assumption that doing 1024 is OK.

I think at a minimum it should lower the limit significantly.

Another option is to only do up to 3 queries, the original, all underscore, and all dashes.

**Comment by epage** [comment]
> Another option is to only do up to 3 queries, the original, all underscore, and all dashes.

I think this is personally reasonable.  I was surprised when I saw what lengths it went through.

**Comment by ehuss** [comment]
Here's a list of names on crates.io that use a mix of dash and underscore. There are 623 entries. That's not as many as I feared.

```
secret-keeper-test_util
sea-strum_macros
serde_sane-rs
server-sent_events
vss_uuid-sys
unic-ucd-name_aliases
va_list-rs
va_list-helper
va_list-test
screenruster-saver-laughing_man
plonk-bls12_381
persian-rug_derive
user_doc-doc_data
user_doc-tests
user_doc-doc_proc_macro
zenoh-link-unixsock_stream
sqlx-core_wasi
sqlx-macros_wasi
sqlx-rt_wasi
a42-__-
msv1_0-sys
msiz_rustc-ap-syntax_pos
msiz_rustc-ap-arena
msiz_rustc-ap-rustc_data_structures
msiz_rustc-ap-rustc_errors
msiz_rustc-ap-graphviz
msiz_rustc-ap-rustc_macros
msiz_rustc-ap-syntax
msiz_rustc-ap-rustc_lexer
msiz_rustc-ap-serialize
msiz_rustc-ap-rustc_target
displaydoc-lite-proc_macros
byte-strings-proc_macros
hkalbasi-rustc-ap-rustc_serialize
hkalbasi-rustc-ap-rustc_graphviz
hkalbasi-rustc-ap-rustc_data_structures
hkalbasi-rustc-ap-rustc_abi
hkalbasi-rustc-ap-rustc_index
hkalbasi-rustc-ap-rustc_macros
google-file1_beta1
google-container1_beta1
google-firebasestorage1_beta-cli
google-billingbudgets1_beta1
google-replicapool1_beta2
google-appengine1_beta5-cli
google-firestore1_beta1
google-datalabeling1_beta1-cli
google-alertcenter1_beta1-cli
google-proximitybeacon1_beta1-cli
google-networkconnectivity1_alpha1
google-appengine1_beta4-cli
google-genomics1_beta2
google-spectrum1_explorer
google-freebase1_sandbox
google-dialogflow2_beta1
google-lifesciences2_beta
google-servicedirectory1_beta1
google-dialogflow2_beta1-cli
google-privateca1_beta1
google-clouduseraccountsvm_beta
google-cloudmonitoring2_beta2-cli
google-datastore1_beta3-cli
google-ml1_beta1-cli
google-recommendationengine1_beta1
google-sasportal1_alpha1-cli
google-ideahub1_beta-cli
google-clouderrorreporting1_beta1
google-sqladmin1_beta4-cli
google-artifactregistry1_beta1
google-firebasedatabase1_beta-cli
google-iap1_beta1
google-area120tables1_alpha1-cli
google-resourceviews1_beta2-cli
google-datalabeling1_beta1
google-autoscaler1_beta2
google-gmailpostmastertools1_beta1-cli
google-cloudkms1_beta1-cli
google-firebasestorage1_beta
google-reseller1_sandbox-cli
google-firestore1_beta1-cli
google-cloudtasks2_beta2-cli
google-transcoder1_beta1-cli
google-cloudprivatecatalog1_beta1-cli
google-clouderrorreporting1_beta1-cli
google-cloudscheduler1_beta1
google-deploymentmanager2_beta1
google-cloudresourcemanager1_beta1
google-datacatalog1_beta1
google-content2_sandbox-cli
google-accesscontextmanager1_beta
google-cloudtasks2_beta3-cli
google-replicapoolupdater1_beta1-cli
google-memcache1_beta2-cli
google-runtimeconfig1_beta1-cli
google-cloudprivatecatalog1_beta1
google-analyticsadmin1_alpha
google-sqladmin1_beta4
google-dns1_beta1
google-gamesconfiguration1_configuration
google-tpu1_alpha1-cli
google-containeranalysis1_beta1
google-admin1_reports
google-oslogin1_beta
google-admin2_email_migration
google-iap1_beta1-cli
google-videointelligence1_beta1-cli
google-cloudprivatecatalogproducer1_beta1
google-reseller1_sandbox
google-analyticsadmin1_alpha-cli
google-analyticsdata1_beta
google-videointelligence1_beta1
google-deploymentmanager2_beta2-cli
google-containeranalysis1_beta1-cli
google-accesscontextmanager1_beta-cli
google-privateca1_beta1-cli
google-gamesconfiguration1_configuration-cli
google-recommendationengine1_beta1-cli
google-fcmdata1_beta1-cli
google-proximitybeacon1_beta1
google-appengine1_beta4
google-datacatalog1_beta1-cli
google-factchecktools1_alpha1
google-resourceviews1_beta2
google-logging2_beta1
google-area120tables1_alpha1
google-firebasehosting1_beta1-cli
google-dlp2_beta1-cli
google-oslogin1_beta-cli
google-logging2_beta1-cli
google-appengine1_beta5
google-cloudasset1_beta1
google-oauth2_v2
google-autoscaler1_beta2-cli
google-commentanalyzer1_alpha1-cli
google-replicapoolupdater1_beta1
google-secretmanager1_beta1-cli
google-cloudsupport2_beta-cli
google-lifesciences2_beta-cli
google-accessapproval1_beta1-cli
google-datafusion1_beta1-cli
google-ml1_beta1
google-accessapproval1_beta1
google-recommender1_beta1-cli
google-gamesmanagement1_management-cli
google-artifactregistry1_beta1-cli
google-metastore1_beta-cli
google-sql1_beta4-cli
google-language1_beta1-cli
google-speech1_beta1
google-cloudmonitoring2_beta2
google-cloudkms1_beta1
google-commentanalyzer1_alpha1
google-binaryauthorization1_beta1-cli
google-cloudsupport2_beta
google-cloudasset1_beta1-cli
google-dataflow1_b4
google-healthcare1_beta1
google-fcmdata1_beta1
google-taskqueue1_beta2-cli
google-runtimeconfig1_beta1
google-firebaseappcheck1_beta-cli
google-cloudprivatecatalogproducer1_beta1-cli
google-gamesmanagement1_management
google-admin1_directory-cli
google-tpu1_alpha1
google-adexchangebuyer2_v2_beta1
google-documentai1_beta2-cli
google-pubsub1_beta2
google-cloudtasks2_beta2
google-memcache1_beta2
google-gmailpostmastertools1_beta1
google-firebasedatabase1_beta
google-language1_beta1
google-billingbudgets1_beta1-cli
google-cloudtasks2_beta3
google-servicedirectory1_beta1-cli
google-logging1_beta3
google-gan1_beta1
google-domains1_beta1-cli
google-taskqueue1_beta2
google-deploymentmanager2_beta2
google-bigqueryconnection1_beta1
google-factchecktools1_alpha1-cli
google-sql1_beta4
google-ideahub1_beta
google-admin1_directory
google-healthcare1_beta1-cli
google-dlp2_beta1
google-datafusion1_beta1
google-file1_beta1-cli
google-alertcenter1_beta1
google-secretmanager1_beta1
google-content2_sandbox
google-recommender1_beta1
google-admin1_reports-cli
google-spectrum1_explorer-cli
google-manager1_beta2-cli
google-prod_tt_sasportal1_alpha1-cli
google-binaryauthorization1_beta1
google-metastore1_beta
google-clouduseraccountsvm_beta-cli
google-cloudscheduler1_beta1-cli
google-sasportal1_alpha1
google-firebase1_beta1
google-pubsub1_beta2-cli
google-gan1_beta1-cli
google-prod_tt_sasportal1_alpha1
google-adexchangebuyer2_v2_beta1-cli
google-networkconnectivity1_alpha1-cli
google-analyticsdata1_beta-cli
google-cloudresourcemanager1_beta1-cli
google-bigqueryconnection1_beta1-cli
google-manager1_beta2
google-speech1_beta1-cli
google-datastore1_beta3
google-domains1_beta1
google-transcoder1_beta1
google-firebasehosting1_beta1
google-documentai1_beta2
google-datastore1_beta2
google-firebase1_beta1-cli
google-replicapool1_beta2-cli
google-firebaseappcheck1_beta
gothack-future-parking_lot
ghosts-proc_macros
dusk-bls12_381-sign
dusk-bls12_381
black_marlin-compiler
black_marlin-macros
blake2-rfc_bellman_edition
blizzard-engine_derive
bls_sigs_ref-rs
blip_buf-sys
mount-ctrl_sampling1000-damping002
mount-ctrl_sampling8000-damping0005
mock-it_codegen
advancedresearch-rigid_body
advancedresearch-graph_builder
advancedresearch-agent_safety_layers
advancedresearch-higher_order_core
advancedresearch-utility_programming
advancedresearch-max_tree
advancedresearch-error_predictive_learning
advancedresearch-tree_mem_sort
advancedresearch-asi_core0
advancedresearch-higher_order_point
advancedresearch-path_iter
advancedresearch-nano_ecs
nix-ptsname_r-shim
gt-bijective_connection_graph
mfplat_vista-sys
mf_vista-sys
nnye-user_server-gateway-business
bech32-no_std
macro_rules_attribute-proc_macro
matio-rs_derive
ntstc_libcmt-sys
ntstc_msvcrt-sys
nt-list_macros
dash-and_underscore
d-_-b
aws-codebuild-status_aws
aws-codebuild-status_terminal
aws-codebuild-status_derive
aws-codebuild-status_server
aws-codebuild-status_web
ggez-assets_manager
do-not-use-this-hello_cratesio
download_rs-silent
hvcg_governance_openapi_catholic-polity
nougat-proc_macros
noise_search_deps_librocksdb-sys
graphannis-malloc_size_of
graphannis-malloc_size_of_derive
allegro_ttf-sys
allegro_audio-sys
allegro_color-sys
allegro_memfile-sys
allegro_image-sys
allegro_primitives-sys
allegro_acodec-sys
allegro_dialog-sys
allegro_font-sys
guessing_game-79888c11-a5a6-4548-a26a-c85ea39d0499
guessing_game-xfossdotcom
guessing_game-panzerstadt
dscfg-cached_file_storage
navinfo-add_two
nacos-api_macro
nameless-clap_generate
nameless-clap_up
nameless-clap_derive
mini_paste-proc_macro
minimult_cortex-m
minimp3_ex-sys
mincore_downlevel-sys
jsonpath_lib-fl
object_store-fork
optimisticpeach-opengles_graphics
inside-vm_arch_support
cursive_buffered_backend-git-branchless
custom-try-proc_macros
jh-x86_64
ignite-rs_derive
crokey-proc_macros
crevice_notan-derive
crate-_-name
crates_io-duplicates
crossbeam-skiplist_piedb
create-rust-app_cli
evr_vista-sys
event_store-core
evercrypt_tiny-sys
const_env_impl--value
const_env--value
core_rustc-serialize
iter-identify_first_last
exonum_libsodium-sys
exonum_librocksdb-sys
ext-trait-proc_macros
cargo-wasi-exe-x86_64-pc-windows-msvc
cargo-wasi-exe-x86_64-apple-darwin
cargo-generate-license_impl-license
cargo-arch_test
cargo_crates-io_docs-rs_test
cargo_crates-io_docs-rs_test2
cargo-wasi-exe-x86_64-unknown-linux-musl
lens-rs_generator
lens-rs_derive
lex-map_editor_std
lending-iterator-proc_macros
lettre-openssl111_email
istherepie-hello_world
lpc177x_8x-hal
jit_rs-sys
flexible-locks_derive
xinput9_1_0-sys
rustc-ap-proc_macro
rustc_codegen_spirv-types
rustc-ap-rustc_errors
rustc-ap-rustc_error_codes
rustc-ap-rustc_feature
rustc-ap-rustc_session
rustc-ap-rustc_ast
rustc-ap-rustc_ast_pretty
rustc-ap-rustc_serialize
rustdt-json_rpc
rustc-ap-rustc_graphviz
rustfmt-config_proc_macro
rustc-ap-rustc_macros
rustc-ap-rustc_parse
rustc-ap-rustc_cratesio_shim
rustc-ap-rustc_lexer
rust_sodium-sys
rustc-ap-rustc_data_structures
rustc-ap-rustc_index
rustc-ap-syntax_pos
rustc-ap-rustc_target
rustc-ap-rustc_ast_passes
rustc-ap-rustc_fs_util
rust-riemann_health
rustc-ap-rustc_expand
rustc-ap-rustc_lint_defs
rust_sodium_holochain_fork-sys
rustc-ap-rustc_span
rustc-ap-rustc_attr
rustc-ap-rustc_arena
rumblebars-rustlex_codegen
tls-client_hello-parser
twain_32-sys
rime_api-sys
webview-sys_suppress
web-view_suppress
rs-tiled_json
rs-odbc_derive
rsudo-rs_support
rs-data-formats_derive
tdameritrade_rust-async
wlc-with_elogind
wlc-sys-with_elogind
quine-mc_cluskey
rhizome_proc-macro-definitions
x86_64-xsave
x86_64-linux-nolibc
ra_ap_la-arena
ra_ap_proc-macro-srv-cli
ra_ap_lsp-server
ra_ap_vfs-notify
ra_ap_rust-analyzer
struct_deser-derive
stm32g473-hal_oppe
stream-tls-client_hello-detector
protoc-bin-vendored-linux-ppcle_64
protoc-bin-vendored-linux-x86_64
protoc-bin-vendored-macos-x86_64
protoc-bin-vendored-linux-x86_32
protoc-bin-vendored-linux-aarch_64
soroban-wasmi_core
solana_libra_grpcio-client
solana_libra_crypto-derive
piston-ai_behavior
piston2d-gfx_graphics
pistoncore-event_loop
piston2d-touch_visualizer
piston-graphics_api_version
pistoncore-sdl2_window
pistoncore-glfw_window
piston2d-glium_graphics
pistoncore-glutin_window
piston-shaders_graphics2d
piston3d-gfx_voxel
piston2d-drag_controller
pistoncore-winit_window
piston2d-scroll_controller
piston-dyon_interactive
piston-timer_controller
piston2d-graphics_tree
piston2d-deform_grid
piston-button_controller
piston-split_controller
piston-button_tracker
piston-gfx_texture
piston2d-wgpu_graphics
piston2d-opengl_graphics
piston-fake_dpi
piston-history_tree
sapling-crypto_ce
safer_ffi-proc_macro
ptr_eq-macros
sugiura-hiromichi_gc
sugiura-hiromichi_mylibrary
sugiura-hiromichi_tp
sugiura-hiromichi_dot
sudo_plugin-sys
vds_uuid-sys
pocket_prover-derive
pocket_prover-set
patch-test-crate_121213131213131
paste_rs-cli
paho_mqtt_c-sys
simple_tables-derive
simple-tlv_derive
simple_tables-core
simple_predicates-rs
update-sync_derive
bootloader-x86_64-bios-boot-sector
bootloader-x86_64-bios-common
bootloader-x86_64-bios-stage-2
bootloader-x86_64-bios-stage-4
bootloader-x86_64-bios-stage-3
bootloader-x86_64-common
bootloader-x86_64-uefi
gps-gunnlaug_18
air-interpreter-wasm_new_it_types
air-interpreter-wasm_update_images
air-interpreter-wasm_new_stepper_outcome
air-interpreter-wasm_prerelease
air-interpreter-wasm_fold_iterable
air-interpreter-wasm_ci_fix_version_increment
air-interpreter-wasm_fix
air-interpreter-wasm_wasm_log
air-interpreter-wasm_update_example
air-interpreter-wasm_get_rid_current_peer_id
air-interpreter-wasm_security_tetraplets_reborn
air-interpreter-wasm_wit
air-interpreter-wasm_jvalue_flatten
air-interpreter-wasm_crate_release
air-interpreter-wasm_branch_release
air-interpreter-wasm_logger-i32
d3d10_1-sys
mbnapi_uuid-sys
android_log-sys
android_liblog-sys
android_looper-sys
android_sensor-sys
merge_derive-hashmap
metrics-exporter-http-async_std
metrics-exporter-async_std
bachue-auto_impl
geng-debug_overlay
get_if_addrs-sys
django-query_derive
bsd_auth-sys
mproxy-socket_dispatch
mycodee-project_manager
my_crate_72a9eb54-5002-4540-a6e6-f2620f51b550
m2-ctrl_asm
m2-ctrl_fsm
gluon_language-server
gluon_c-api
aux_ulib-sys
aorist_extendr-macros
aorist_extendr-api
aorist_extendr-engine
dyn_safe-proc_macros
mdbook-pdf-headless_chrome
next-gen-proc_macros
next-gen_proc-macro
a-star_traitbased
bird_tool_utils-man
aflak_imgui-sys
aflak_imgui-glium-renderer
gmt_dos-clients_domeseeing
gmt_mount-ctrl_sampling8000-damping0005
gmt_dos-clients_arrow
gmt_dos-clients_io
gmt_mount-ctrl_sampling8000-damping002
gmt_m1-ctrl_hardpoints-dynamics
gmt_m1-ctrl_actuators
gmt_mount-ctrl_sampling8000-damping002_ze60
gmt_mount-ctrl_sampling1000-damping002
gmt_mount-ctrl_sampling1000-damping002_ze00
gmt_dos-clients_crseo
gmt_dos-actors
gmt_mount-ctrl_sampling1000-damping002_ze30
gmt_m1-ctrl_center-actuators
gmt_mount-ctrl
gmt_mount-ctrl_sampling1000-damping002_ze30-500hz
gmt_mount-ctrl_sampling8000-damping002_ze00
gmt_m1-ctrl_outer-actuators
gmt_dos-clients_mount
gmt_mount-ctrl_sampling8000-damping002_ze30
gmt_m1-ctrl
gmt_mount-ctrl_sampling1000-damping002_ze60
deno-x86_64-pc-windows-msvc
deno-x86_64-apple-darwin
deno-x86_64-unknown-linux-gnu
defile-proc_macros
asteracea_proc-macro-definitions
async_ach-pubsub
async_ach-ring
async_ach-watch
async_fn-proc_macros
async_ach-cell
async_ach-notify
async_ach-mpmc
async_ach-waker
async_ach-spsc
asmkit-x86_64
azul-wr_malloc_size_of
gdnative-visual_script
konektor_db-sys
koibumi-unifont_jp
fix-hidden-lifetime-bug-proc_macros
jelly-mem_access
lukas_zech_software_hello-world
fruit-salad_proc-macro-definitions
fuel-pest_meta
fuel-pest_derive
fuel-pest_generator
fuel-pest_grammars
fullcodec-bls12_381
function_name-proc-macro
future-parking_lot
eskom_se_push-api
jpark011-guessing_game
libpg_query-sys
libpg_query2-sys
lim1988_cargo-demo
linux_input-sys
limn-text_layout
liquidcrystal_i2c-rs
lib_remotebuild-rs
libaeron_driver-sys
libzfs_core-sys
encoding-next_index_tests
entity_data-macros
fawkes-crypto-pairing_ce
fawkes-crypto-bellman_ce
fawkes-crypto_derive
fast-rustc-ap-rustc_data_structures
fast-rustc-ap-rustc_ast
fast-rustc-ap-rustc_macros
fast-rustc-ap-rustc_index
fast-rustc-ap-rustc_errors
fast-rustc-ap-rustc_feature
fast-rustc-ap-rustc_ast_pretty
fast-rustc-ap-rustc_session
fast-rustc-ap-rustc_attr
fast-rustc-ap-rustc_span
fast-rustc-ap-rustc_parse
fast-rustc-ap-rustc_lexer
fast-rustc-ap-rustc_target
fast-rustc-ap-rustc_fs_util
ff-uint_derive
ff_derive-zeroize
foley-cargoexample-guessing_game
lz4_flex-util
egui-latest_wgpu_backend
clap_derive-v3
clean-node_modules
cloudflare-dns-updater_rs
erlang_nif-sys
login_cap-sys
ip_network_table-deps-treebitmap
lazy-regex-proc_macros
lattice_qcd_rs-procedural_macro
qemu-x86_64
xmltree-parse_with_config
qpid_proton-sys
winapi-x86_64-pc-windows-gnu
with_builtin_macros-proc_macros
with_locals-proc_macros
tamasfe-schemars_derive
thiserror_core2-impl
rv32m1_ri5cy-hal
rv32m1_ri5cy-pac
to-syn-value_derive
toy-arms_derive
toml-query_derive
tokio-stream_wasi
tokio_modbus-winets
tokio-util_wasi
tokio-tungstenite_wasi
tokio-postgres_wasi
ws2_32-sys
wsbapp_uuid-sys
rest-client_codegen
retryable-proc_macros
tibco_ems-sys
rpi_ws281x-c
rpi_ws281x-sys
r2d2-influx_db_client
```

**Comment by arlosi** [comment]
I did some more analysis on the names on crates.io. The currently limit of 1024 is definitely excessive, as nothing uses more than 8 dashes/underscores.

We'd need to set the limit to at least 16 in order to have fewer misses than the suggested algorithm of (original, underscores, dashes).

| Request limit | Potentially missed crates |
|-------|--------|
| 1     | 68282  |
| 2     | 21214  |
| 4     | 4437   |
| 8     | 897    |
| 16    | 177    |
| 32    | 37     |
| 64    | 12     |
| 128   | 2      |
| 256   | 0      |

**Comment by est31** [comment]
cc #2775 and https://internals.rust-lang.org/t/pre-rfc-unify-dashes-and-underscores-on-crates-io/13216/9

**Comment by est31** [comment]
What about doing the requests in parallel? This is also how `cargo update` works, and a `cargo update` to a normal-sized repository will create requests to a couple hundred crates, so number wise it's not that excessive (`cat Cargo.lock | rg "^name =" | sort | uniq | wc -l` gives me 279 for the cargo repo for example). The problem might be the high ratio of 404s returned though, maybe some cloud providers don't like that.

**Comment by arlosi** [comment]
> What about doing the requests in parallel?

Cargo does issue the requests in parallel. The motivation for investigating this was the high number of 404s that infra was seeing in logs.

**Comment by est31** [comment]
@arlosi I see. Is there some public discussion about the 404 issue?

I suppose if only 623 entries are affected then it's okay to cut some corners and emit suggestion-free diagnostics in those cases, in all other cases the - and _ only queries will show up the right result.

Ideally one would have used the migration to http indices to migrate to using the canonicalized name (as [suggested in 2020](https://internals.rust-lang.org/t/pre-rfc-unify-dashes-and-underscores-on-crates-io/13216/20)), but I guess now it's too late for that.

**Comment by arlosi** [comment]
> Is there some public discussion about the 404 issue?

Yes, on the [crates-io zulip](https://rust-lang.zulipchat.com/#narrow/stream/318791-t-crates-io/topic/4xx.20rate.20cloudfront.20investigation).

**Comment by Eh2406** [comment]
Sparse registries have the same format as a checkout of the git index. You can transition by checking out the git index and running a static file server in the resulting folder. That transition path was important enough to me to justify not fixing known problems with the index format. Feel free to blame me for that decision, especially as the decision was made knowingly.

Another decision that complicates things is that alternative registries are allowed to have multiple packages whose names only differ by `-`/`_`. Thanks to package renaming, you can even use both packages in the same project.

Thinking about it more, the currently stable implementation tries all kinds of files and returns the first one it finds. (All kinds of files include ridiculously incorrect things like `a_/b-/a-b_foo` as fixed in #11936.) It then reads the available packages from that first file. Rows whose `name` matches the value requested by the dependency get tried by the resolver, and if there are none the names that do not match are included as a suggestion. If I am correct, a registry could have multiple crates that differ only in `-`/`_` with them all listed in the same file in the index. It would work correctly, ignoring the inefficient implementation, even if that one file wast at something hilariously invalid.

Assuming all of this is correct, I suggest that:
1. Cargo be changed to only try the name as given, followed by the name canonicalized to `_`.
1. The registry documentation should be updated to recommend that:
 - If the registry mainly supports cargoes after the change from 1., keeping one index file (canonicalized to `_`) with rows for all packages whose name canonical eyes to the same thing.
 - If the registry needs to support cargoes from before the un-canonical loop was added, then include a file per package based on its name.
 - Explain that  in between "the un-canonical loop was added" and "change from 1." Cargo will get the correct answer if the new system is used but may not do so efficiently.
3. Update the crates.io index to have both `a-/b_/a-b_foo` and `a_/b_/a_b_foo` files in both the git and sparse index.
4. (If the crates.io git and sparse indexes are allowed to differ, and we have #10928) consider removing `a-/b_/a-b_foo` from the sparse and removing `a_/b_/a_b_foo` from the git.

**Comment by weihanglo** [comment]
I am going to close this since #11937 and #12083 significantly reduces network requests. If there is something still missing after #12083, feel free to reopen :)
